mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/RequestState.scala

```diff
@@ -375,22 +376,25 @@ import scala.util.{Failure, Success}
 
   // Processing
 
-  def main(registrantsByPacketId: Map[PacketId, ActorRef[A]], nextPacketId: PacketId): Behavior[Request[A]] =
+  def main(registrantsByPacketId: Map[PacketId, ActorRef[A]], lastPacketId: Option[PacketId]): Behavior[Request[A]] =
     Behaviors.receiveMessage {
-      case Register(registrant: ActorRef[A], reply) if nextPacketId.underlying <= MaxPacketId.underlying =>
+      case Register(registrant: ActorRef[A], reply) if registrantsByPacketId.size < MaxNrOfPacketId =>
+        @tailrec
+        def step(id: PacketId): PacketId =
+          if (id.underlying == MaxPacketId.underlying)
+            step(MinPacketId)
+          else if (registrantsByPacketId.contains(id))
+            step(PacketId(id.underlying + 1))
+          else
+            id
+        val nextPacketId = step(lastPacketId.fold(MinPacketId)(l => PacketId(l.underlying + 1)))
         reply.success(Registered(nextPacketId))
-        main(registrantsByPacketId + (nextPacketId -> registrant), PacketId(nextPacketId.underlying + 1))
+        main(registrantsByPacketId + (nextPacketId -> registrant), Some(nextPacketId))
       case _: Register[A] =>
         Behaviors.same // We cannot allocate any more. This will eventually cause a timeout to occur on the requestor.
```

Comment:
Note that this means the `Promise` will never be completed, so that needs to be dealt with here. Additionally, any remaining `Promise`s need to be failed when this actor stops.

In general, I don't see the need to fail in this condition. I think it's better to track the registrants and fulfill the promise on the next `Unregister`. I'd say that's the main difference between the two PRs.